### ISSUE TITLE
Acting on non-rectangular selections

### ISSUE THREAD

**Issue by diamond-lizard** [noctuid/rect-ext.el#6]
When I use control-v to visually select the following text:
```
111111111111111
11111111111111111111
1111111111
```
and type 
```
:`<,`>B s;1;2;g
```
The result is:
```
222222222211111
22222222221111111111
2222222222
```
It appears that the B function is limited to acting on rectangular selections only.  Which makes sense, as this is `rect-ext.el`, after all, so I'm not sure if this is a bug report so much as a feature request.

It'd be great if B could work on every type of visual selection, not just rectangles.  Is this possible?

**Comment by noctuid** [comment]
Right, currently `B` is for block selections only. It would be better to just use `s` here right? What would be the use case of `B` outside of visual block selections (just consistency)? 

Part of the reason I haven't worked more on this is that I don't know where I want to take this package (e.g. focusing on rectangles vs. narrowing; there are a lot of alternative narrowing packages already) and `B` with `s` isn't  nearly as nice as `s` alone becuase of the visualization `s` does. Trying to add that would be a lot more work than went into the package up to this point. There's also the alternative of potentially getting visual-block support baked into these commands, so you could just do ``:`<,`>s`` (this will already work for several evil commands)  or make it look like that but still use a generic wrapper. I'd like to get rid of `B` at some point, but for now, it wouldn't be too hard to have `B` work with other selections if you think it would be useful.

**Comment by diamond-lizard** [comment]
Of course in the example I gave above it would make much more sense to just use `:s`

Here's a more real-world example that demonstrates where I'd find `B` useful:

I have a buffer that looks like:
```
111111111111111            111111111111111
11111111111111111111       11111111111111111111
1111111111                 1111111111
```

and I need to make substitutions on just the right half of this text.

Here a simple `:s` would not do, as it would substitute everywhere on the lines, while I want to restrict the subsitution to just the right half.

And, of course, limiting the substitution area to just a rectangle won't be sufficient either, for the reason described above.
```
:`<,`>s;1;2;g
```
with the right half of that text selected doesn't work either.  I really need `B` to work with the entire right half, which is how it works in vim.

**Comment by diamond-lizard** [comment]
Regarding getting rid of `B`, I'd be all for it (pardon the pun), if `:s` would work the way I need it to on partial-line visual selections.

**Comment by noctuid** [comment]
If I understand correctly, you still want a block selection, you just want it to cover all the 1s on the right. That's something I'll have to think more about. Emacs has various different types of rectangle select modes builtin and various ways to get around this issue. For example, `rectangle-mark-mode` (builtin), `phi-rectangle-mode` (not builtin), etc. will automatically temporarily add spaces to the shorter lines when you are selecting (only the last one matters though). There is also `picture-mode` which allows for this. The issue is that none of these seem to play well with evil, and evil's visual block mode does not try to mimic this functionality (i.e. it has no equivalent of vim's `virtualedit=block`).

To handle this case automatically when you select your example would require more fundamental/complicated changes (to evil).

`B` already works with non-visual-block selections. My workaround suggestion for now would probably be to use the `rectangle-utils` (not builtin) package's `rectangle-utils-extend-rectangle-to-end` command in visual-char (not block) mode, which will add the necessary space padding.
